File: apps/web/lib/privacy/pii.ts
export type PiiRedactionKind =
  | "credit_card"
  | "email"
  | "iban"
  | "ip_address"
  | "national_id"
  | "phone"
  | "secret";

export type PiiRedactionCounts = Partial<Record<PiiRedactionKind, number>>;

export type PiiSanitizationResult<T> = {
  value: T;
  redactions: PiiRedactionCounts;
  redacted: boolean;
};

// Secrets are destructively redacted and never rehydrated: a credential must
// not be able to round-trip through a (possibly prompt-injected) model back
// into generated schemas or outputs. All other kinds are reversible workflow
// data — the model sees a stable numbered placeholder and the real value is
// substituted back on our side (see PseudonymizationSession).
export const SECRET_PLACEHOLDER = "[REDACTED_SECRET]";

const REVERSIBLE_KINDS = [
  "email",
  "iban",
  "national_id",
  "ip_address",
  "phone",
  "credit_card",
] as const;

type ReversibleKind = (typeof REVERSIBLE_KINDS)[number];

const SECRET_ASSIGNMENT_REGEX =
  /\b((?:api[_-]?key|secret|token|access[_-]?token|refresh[_-]?token|authorization)\s*[:=]\s*["']?)([A-Za-z0-9._~+/=-]{12,})(["']?)/gi;
const BEARER_TOKEN_REGEX = /\b(Authorization\s*:\s*Bearer\s+)([A-Za-z0-9._~+/=-]{12,})/gi;
const PREFIXED_SECRET_REGEX =
  /\b(?:sk|pk|rk|ghp|gho|ghu|ghs|glpat|github_pat|xox[baprs]?)[-_][A-Za-z0-9_=-]{12,}\b/gi;
const EMAIL_REGEX = /\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b/gi;
const IBAN_REGEX = /\b[A-Z]{2}\d{2}(?:[ ]?[A-Z0-9]){11,30}\b/gi;
const CREDIT_CARD_CANDIDATE_REGEX = /\b(?:\d[ -]*?){13,19}\b/g;
const NATIONAL_ID_REGEX = /\b\d{3}-\d{2}-\d{4}\b/g;
const IPV4_REGEX =
  /\b(?:(?:25[0-5]|2[0-4]\d|1?\d?\d)\.){3}(?:25[0-5]|2[0-4]\d|1?\d?\d)\b/g;
const INTERNATIONAL_PHONE_REGEX = /(?:\+|00)\d[\d\s().-]{7,}\d/g;
const US_PHONE_REGEX = /(?:\(\d{3}\)|\b\d{3})[\s.-]\d{3}[\s.-]\d{4}\b/g;

// Matches any placeholder this module can emit — regex kinds ([EMAIL_3]) and
// registered known-value categories ([GMAIL_LABEL_2]). Rehydration only
// substitutes placeholders the session actually issued (map lookup), so the
// broader pattern cannot leak anything.
const PLACEHOLDER_REGEX = /\[([A-Z][A-Z0-9_]*)_(\d+)\]/g;

// Known values shorter than this are too likely to collide with ordinary
// words ("To", "Do") to substitute in free text. They still get a placeholder
// for the capability listing and rehydration.
const MIN_KNOWN_VALUE_SUBSTITUTION_LENGTH = 3;

function addCount(redactions: PiiRedactionCounts, kind: PiiRedactionKind, count: number) {
  if (count > 0) redactions[kind] = (redactions[kind] ?? 0) + count;
}

function luhnCheck(digits: string) {
  let sum = 0;
  let shouldDouble = false;

  for (let i = digits.length - 1; i >= 0; i -= 1) {
    let digit = Number(digits[i]);
    if (Number.isNaN(digit)) return false;

    if (shouldDouble) {
      digit *= 2;
      if (digit > 9) digit -= 9;
    }

    sum += digit;
    shouldDouble = !shouldDouble;
  }

  return sum > 0 && sum % 10 === 0;
}

export function mergePiiRedactions(...items: Array<PiiRedactionCounts | null | undefined>): PiiRedactionCounts {
  const merged: PiiRedactionCounts = {};

  for (const item of items) {
    if (!item) continue;
    for (const [kind, count] of Object.entries(item) as Array<[PiiRedactionKind, number | undefined]>) {
      addCount(merged, kind, count ?? 0);
    }
  }

  return merged;
}

export function hasPiiRedactions(redactions: PiiRedactionCounts) {
  return Object.values(redactions).some((count) => (count ?? 0) > 0);
}

/**
 * Reversible, request-scoped PII pseudonymization.
 *
 * Each unique raw value gets one stable numbered placeholder for the lifetime
 * of the session ("max@firma.de" → "[EMAIL_1]" everywhere), so the model can
 * reason about identity across prompt parts. The mapping lives only in this
 * object — it is never persisted and never leaves the server — and
 * `rehydrate*` substitutes the real values back into model output (e.g. a
 * generated workflow schema) before it is used.
 *
 * Secrets are the exception: destructively redacted, never mapped.
 */
export class PseudonymizationSession {
  private placeholderByValue = new Map<string, string>();
  private valueByPlaceholder = new Map<string, string>();
  private counters = new Map<string, number>();
  // Known values registered via registerKnownValue, for applyKnownValues.
  private knownValues: Array<{ raw: string; placeholder: string }> = [];

  private placeholderFor(kind: ReversibleKind, raw: string): string {
    const key = `${kind}\x00${raw}`;
    const existing = this.placeholderByValue.get(key);
    if (existing !== undefined) return existing;
    const number = (this.counters.get(kind) ?? 0) + 1;
    this.counters.set(kind, number);
    const placeholder = `[${kind.toUpperCase()}_${number}]`;
    this.placeholderByValue.set(key, placeholder);
    this.valueByPlaceholder.set(placeholder, raw);
    return placeholder;
  }

  private subReversible(text: string, kind: ReversibleKind, regex: RegExp) {
    let count = 0;
    const value = text.replace(regex, (match) => {
      count += 1;
      return this.placeholderFor(kind, match);
    });
    return { value, count };
  }

  private subSecretAssignments(text: string) {
    let count = 0;
    const assigned = text.replace(SECRET_ASSIGNMENT_REGEX, (_match, prefix: string, _secret: string, suffix: string) => {
      count += 1;
      return `${prefix}${SECRET_PLACEHOLDER}${suffix}`;
    });
    const bearer = assigned.replace(BEARER_TOKEN_REGEX, (_match, prefix: string) => {
      count += 1;
      return `${prefix}${SECRET_PLACEHOLDER}`;
    });

    return { value: bearer, count };
  }

  private subCreditCards(text: string) {
    let count = 0;
    const value = text.replace(CREDIT_CARD_CANDIDATE_REGEX, (match) => {
      const digits = match.replace(/\D/g, "");
      if (digits.length < 13 || digits.length > 19 || !luhnCheck(digits)) {
        return match;
      }

      count += 1;
      return this.placeholderFor("credit_card", match);
    });

    return { value, count };
  }

  sanitizeText(input: string): PiiSanitizationResult<string> {
    let value = input;
    const redactions: PiiRedactionCounts = {};

    const secretAssignments = this.subSecretAssignments(value);
    value = secretAssignments.value;
    addCount(redactions, "secret", secretAssignments.count);

    let prefixedCount = 0;
    value = value.replace(PREFIXED_SECRET_REGEX, () => {
      prefixedCount += 1;
      return SECRET_PLACEHOLDER;
    });
    addCount(redactions, "secret", prefixedCount);

    for (const [kind, regex] of [
      ["email", EMAIL_REGEX],
      ["iban", IBAN_REGEX],
      ["national_id", NATIONAL_ID_REGEX],
      ["ip_address", IPV4_REGEX],
      ["phone", INTERNATIONAL_PHONE_REGEX],
      ["phone", US_PHONE_REGEX],
    ] as const) {
      const result = this.subReversible(value, kind as ReversibleKind, regex);
      value = result.value;
      addCount(redactions, kind, result.count);
    }

    const creditCards = this.subCreditCards(value);
    value = creditCards.value;
    addCount(redactions, "credit_card", creditCards.count);

    return {
      value,
      redactions,
      redacted: hasPiiRedactions(redactions),
    };
  }

  private sanitizeUnknown(value: unknown, seen: WeakSet<object>): PiiSanitizationResult<unknown> {
    if (typeof value === "string") return this.sanitizeText(value);
    if (value === null || typeof value !== "object") {
      return { value, redactions: {}, redacted: false };
    }

    if (seen.has(value)) {
      return { value: "[REDACTED_CIRCULAR_REFERENCE]", redactions: {}, redacted: false };
    }
    seen.add(value);

    if (Array.isArray(value)) {
      const redactions: PiiRedactionCounts = {};
      const sanitized = value.map((item) => {
        const result = this.sanitizeUnknown(item, seen);
        Object.assign(redactions, mergePiiRedactions(redactions, result.redactions));
        return result.value;
      });

      seen.delete(value);
      return {
        value: sanitized,
        redactions,
        redacted: hasPiiRedactions(redactions),
      };
    }

    const redactions: PiiRedactionCounts = {};
    const sanitized: Record<string, unknown> = {};

    for (const [key, item] of Object.entries(value)) {
      const sanitizedKey = this.sanitizeText(key);
      const sanitizedValue = this.sanitizeUnknown(item, seen);
      const nextRedactions = mergePiiRedactions(redactions, sanitizedKey.redactions, sanitizedValue.redactions);
      Object.assign(redactions, nextRedactions);
      sanitized[sanitizedKey.value] = sanitizedValue.value;
    }

    seen.delete(value);
    return {
      value: sanitized,
      redactions,
      redacted: hasPiiRedactions(redactions),
    };
  }

  sanitizeValue<T>(value: T): PiiSanitizationResult<T> {
    const result = this.sanitizeUnknown(value, new WeakSet<object>());
    return {
      value: result.value as T,
      redactions: result.redactions,
      redacted: result.redacted,
    };
  }

  /**
   * Register a value known to be user-created (e.g. a Gmail label name or a
   * Notion property name discovered via connection introspection) under a
   * category like "GMAIL_LABEL". Returns the stable placeholder for it — the
   * same raw value in the same category always maps to the same placeholder,
   * and rehydration substitutes it back exactly like regex-detected PII.
   */
  registerKnownValue(category: string, raw: string): string {
    const normalized = category.toUpperCase().replace(/[^A-Z0-9]+/g, "_").replace(/^_+|_+$/g, "");
    if (!/^[A-Z][A-Z0-9_]*$/.test(normalized)) {
      throw new Error(`Invalid known-value category: ${category}`);
    }
    const key = `${normalized}\x00${raw}`;
    const existing = this.placeholderByValue.get(key);
    if (existing !== undefined) return existing;
    const number = (this.counters.get(normalized) ?? 0) + 1;
    this.counters.set(normalized, number);
    const placeholder = `[${normalized}_${number}]`;
    this.placeholderByValue.set(key, placeholder);
    this.valueByPlaceholder.set(placeholder, raw);
    this.knownValues.push({ raw, placeholder });
    return placeholder;
  }

  /**
   * Replace occurrences of registered known values in free text with their
   * placeholders, so text the user typed ("label it Invoices") aligns with the
   * placeholder identity used in capability listings ([GMAIL_LABEL_1]).
   * Longest-first so overlapping names ("Sales" / "Sales EU") resolve to the
   * more specific match; case-insensitive because users rarely type exact case.
   */
  applyKnownValues(text: string): string {
    if (this.knownValues.length === 0) return text;
    let value = text;
    const candidates = this.knownValues
      .filter((entry) => entry.raw.length >= MIN_KNOWN_VALUE_SUBSTITUTION_LENGTH)
      .sort((a, b) => b.raw.length - a.raw.length);
    for (const { raw, placeholder } of candidates) {
      const escaped = raw.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
      // Word-ish boundaries only where the value starts/ends with a word char,
      // so names like "#sales-data" still match after a space. Underscore is
      // part of the boundary class so a name like "status" can never match
      // inside an already-substituted placeholder such as [NOTION_STATUS_1].
      const prefix = /^[A-Za-z0-9]/.test(raw) ? "(?<![A-Za-z0-9_])" : "";
      const suffix = /[A-Za-z0-9]$/.test(raw) ? "(?![A-Za-z0-9_])" : "";
      value = value.replace(new RegExp(`${prefix}${escaped}${suffix}`, "gi"), placeholder);
    }
    return value;
  }

  /** applyKnownValues over every string in a nested structure (values and keys). */
  applyKnownValuesToValue<T>(value: T): T {
    if (this.knownValues.length === 0) return value;
    return this.applyKnownValuesUnknown(value, new WeakSet<object>()) as T;
  }

  private applyKnownValuesUnknown(value: unknown, seen: WeakSet<object>): unknown {
    if (typeof value === "string") return this.applyKnownValues(value);
    if (value === null || typeof value !== "object") return value;
    if (seen.has(value)) return value;
    seen.add(value);

    try {
      if (Array.isArray(value)) {
        return value.map((item) => this.applyKnownValuesUnknown(item, seen));
      }
      const result: Record<string, unknown> = {};
      for (const [key, item] of Object.entries(value)) {
        result[this.applyKnownValues(key)] = this.applyKnownValuesUnknown(item, seen);
      }
      return result;
    } finally {
      seen.delete(value);
    }
  }

  /**
   * Substitute known placeholders back to their real values. Placeholders this
   * session never issued (e.g. a model-invented "[EMAIL_99]") are left as-is.
   */
  rehydrateText(text: string): string {
    if (this.valueByPlaceholder.size === 0) return text;
    return text.replace(PLACEHOLDER_REGEX, (match) => this.valueByPlaceholder.get(match) ?? match);
  }

  rehydrateValue<T>(value: T): T {
    return this.rehydrateUnknown(value, new WeakSet<object>()) as T;
  }

  private rehydrateUnknown(value: unknown, seen: WeakSet<object>): unknown {
    if (typeof value === "string") return this.rehydrateText(value);
    if (value === null || typeof value !== "object") return value;
    if (seen.has(value)) return value;
    seen.add(value);

    try {
      if (Array.isArray(value)) {
        return value.map((item) => this.rehydrateUnknown(item, seen));
      }
      const result: Record<string, unknown> = {};
      for (const [key, item] of Object.entries(value)) {
        result[this.rehydrateText(key)] = this.rehydrateUnknown(item, seen);
      }
      return result;
    } finally {
      seen.delete(value);
    }
  }
}

// ── Stateless wrappers (one-shot, mapping discarded → effectively destructive) ──
//
// Kept for callers without a request lifecycle. New code that consumes model
// output should hold a PseudonymizationSession and rehydrate.

export function sanitizeTextForLlm(input: string): PiiSanitizationResult<string> {
  return new PseudonymizationSession().sanitizeText(input);
}

export function sanitizeValueForLlm<T>(value: T): PiiSanitizationResult<T> {
  return new PseudonymizationSession().sanitizeValue(value);
}

File: apps/runtime/engine/pii.py
from __future__ import annotations

from dataclasses import dataclass
import re
from typing import Any, Callable, Optional


PiiRedactionCounts = dict[str, int]


@dataclass(frozen=True)
class PiiSanitizationResult:
    value: Any
    redactions: PiiRedactionCounts

    @property
    def redacted(self) -> bool:
        return any(count > 0 for count in self.redactions.values())


# Secrets are destructively redacted and never rehydrated: a credential must
# not be able to round-trip through a (possibly prompt-injected) model back
# into outputs, messages, or tool arguments. All other kinds are reversible
# workflow data — the LLM sees a stable numbered placeholder and the real
# value is substituted back on our side (see PseudonymizationSession).
SECRET_PLACEHOLDER = "[REDACTED_SECRET]"

REVERSIBLE_KINDS = (
    "email",
    "iban",
    "national_id",
    "ip_address",
    "phone",
    "credit_card",
    # Only emitted when a session is built with a name detector (strict mode).
    "person",
)

SECRET_ASSIGNMENT_RE = re.compile(
    r'\b((?:api[_-]?key|secret|token|access[_-]?token|refresh[_-]?token|authorization)\s*[:=]\s*["\']?)'
    r'([A-Za-z0-9._~+/=-]{12,})(["\']?)',
    re.IGNORECASE,
)
BEARER_TOKEN_RE = re.compile(r"\b(Authorization\s*:\s*Bearer\s+)([A-Za-z0-9._~+/=-]{12,})", re.IGNORECASE)
PREFIXED_SECRET_RE = re.compile(
    r"\b(?:sk|pk|rk|ghp|gho|ghu|ghs|glpat|github_pat|xox[baprs]?)[-_][A-Za-z0-9_=-]{12,}\b",
    re.IGNORECASE,
)
EMAIL_RE = re.compile(r"\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b", re.IGNORECASE)
IBAN_RE = re.compile(r"\b[A-Z]{2}\d{2}(?:[ ]?[A-Z0-9]){11,30}\b", re.IGNORECASE)
CREDIT_CARD_CANDIDATE_RE = re.compile(r"\b(?:\d[ -]*?){13,19}\b")
NATIONAL_ID_RE = re.compile(r"\b\d{3}-\d{2}-\d{4}\b")
IPV4_RE = re.compile(r"\b(?:(?:25[0-5]|2[0-4]\d|1?\d?\d)\.){3}(?:25[0-5]|2[0-4]\d|1?\d?\d)\b")
INTERNATIONAL_PHONE_RE = re.compile(r"(?:\+|00)\d[\d\s().-]{7,}\d")
US_PHONE_RE = re.compile(r"(?:\(\d{3}\)|\b\d{3})[\s.-]\d{3}[\s.-]\d{4}\b")

# Matches placeholders this module emits, e.g. [EMAIL_3] or [CREDIT_CARD_12].
# Tolerant on purpose: model output may attach suffixes ("[EMAIL_1]'s") but the
# bracketed token itself survives generation intact.
_PLACEHOLDER_RE = re.compile(r"\[(" + "|".join(kind.upper() for kind in REVERSIBLE_KINDS) + r")_(\d+)\]")


def _add_count(redactions: PiiRedactionCounts, kind: str, count: int) -> None:
    if count > 0:
        redactions[kind] = redactions.get(kind, 0) + count


def merge_redactions(*items: PiiRedactionCounts | None) -> PiiRedactionCounts:
    merged: PiiRedactionCounts = {}
    for item in items:
        if not item:
            continue
        for kind, count in item.items():
            _add_count(merged, kind, count)
    return merged


def _luhn_check(digits: str) -> bool:
    total = 0
    should_double = False

    for char in reversed(digits):
        if not char.isdigit():
            return False

        digit = int(char)
        if should_double:
            digit *= 2
            if digit > 9:
                digit -= 9

        total += digit
        should_double = not should_double

    return total > 0 and total % 10 == 0


class PseudonymizationSession:
    """Reversible, run-scoped PII pseudonymization.

    Each unique raw value gets one stable numbered placeholder for the whole
    run ("max@firma.de" → "[EMAIL_1]" everywhere), so the model can reason
    about identity across prompts, tool results, and turns. The mapping lives
    only in this object's memory — it is never persisted and never leaves the
    runtime process — and `rehydrate_*` substitutes the real values back into
    model output and tool arguments before they take effect.

    Secrets are the exception: destructively redacted, never mapped.

    With a `name_detector` (strict mode), person names found by local NER are
    pseudonymized as [PERSON_n] too. Detection runs entirely in-process; the
    detector returns the name strings to replace.
    """

    def __init__(self, name_detector: Optional[Callable[[str], list[str]]] = None) -> None:
        self._placeholder_by_value: dict[tuple[str, str], str] = {}
        self._value_by_placeholder: dict[str, str] = {}
        self._counters: dict[str, int] = {}
        self._name_detector = name_detector

    # ── pseudonymize ──────────────────────────────────────────────────────────

    def _placeholder_for(self, kind: str, raw: str) -> str:
        key = (kind, raw)
        existing = self._placeholder_by_value.get(key)
        if existing is not None:
            return existing
        number = self._counters.get(kind, 0) + 1
        self._counters[kind] = number
        placeholder = f"[{kind.upper()}_{number}]"
        self._placeholder_by_value[key] = placeholder
        self._value_by_placeholder[placeholder] = raw
        return placeholder

    def _sub_reversible(self, text: str, kind: str, pattern: re.Pattern[str]) -> tuple[str, int]:
        count = 0

        def replace(match: re.Match[str]) -> str:
            nonlocal count
            count += 1
            return self._placeholder_for(kind, match.group(0))

        return pattern.sub(replace, text), count

    def _sub_secret_assignments(self, text: str) -> tuple[str, int]:
        count = 0

        def replace_assignment(match: re.Match[str]) -> str:
            nonlocal count
            count += 1
            return f"{match.group(1)}{SECRET_PLACEHOLDER}{match.group(3)}"

        def replace_bearer(match: re.Match[str]) -> str:
            nonlocal count
            count += 1
            return f"{match.group(1)}{SECRET_PLACEHOLDER}"

        value = SECRET_ASSIGNMENT_RE.sub(replace_assignment, text)
        value = BEARER_TOKEN_RE.sub(replace_bearer, value)
        return value, count

    def _sub_credit_cards(self, text: str) -> tuple[str, int]:
        count = 0

        def replace_card(match: re.Match[str]) -> str:
            nonlocal count
            digits = re.sub(r"\D", "", match.group(0))
            if len(digits) < 13 or len(digits) > 19 or not _luhn_check(digits):
                return match.group(0)

            count += 1
            return self._placeholder_for("credit_card", match.group(0))

        return CREDIT_CARD_CANDIDATE_RE.sub(replace_card, text), count

    def sanitize_text(self, text: str) -> PiiSanitizationResult:
        value = text
        redactions: PiiRedactionCounts = {}

        value, count = self._sub_secret_assignments(value)
        _add_count(redactions, "secret", count)

        value, count = PREFIXED_SECRET_RE.subn(SECRET_PLACEHOLDER, value)
        _add_count(redactions, "secret", count)

        for kind, pattern in (
            ("email", EMAIL_RE),
            ("iban", IBAN_RE),
            ("national_id", NATIONAL_ID_RE),
            ("ip_address", IPV4_RE),
            ("phone", INTERNATIONAL_PHONE_RE),
            ("phone", US_PHONE_RE),
        ):
            value, count = self._sub_reversible(value, kind, pattern)
            _add_count(redactions, kind, count)

        value, count = self._sub_credit_cards(value)
        _add_count(redactions, "credit_card", count)

        value, count = self._sub_person_names(value)
        _add_count(redactions, "person", count)

        return PiiSanitizationResult(value=value, redactions=redactions)

    def _sub_person_names(self, text: str) -> tuple[str, int]:
        """Strict mode only: replace NER-detected person names with [PERSON_n].

        Runs after the structured patterns so detection sees the already
        pseudonymized text (it can't re-find an email as part of a name).
        Longest names first, so "Max Mustermann" wins over a bare "Max".
        A detector failure degrades to no name replacement — same contract as
        not having a detector at all.
        """
        if self._name_detector is None or not text:
            return text, 0

        try:
            detected = self._name_detector(text)
        except Exception as exc:
            print(f"[pii-ner] name detection failed; continuing without: {exc}", flush=True)
            return text, 0

        names = sorted(
            {name.strip() for name in detected if name and name.strip() and "[" not in name},
            key=len,
            reverse=True,
        )
        if not names:
            return text, 0

        count = 0
        value = text
        for name in names:
            placeholder = None
            pattern = re.compile(rf"(?<!\w){re.escape(name)}(?!\w)")

            def replace(_match: re.Match[str]) -> str:
                nonlocal count, placeholder
                count += 1
                if placeholder is None:
                    placeholder = self._placeholder_for("person", name)
                return placeholder

            value = pattern.sub(replace, value)

        return value, count

    def sanitize_value(self, value: Any, _seen: set[int] | None = None) -> PiiSanitizationResult:
        if isinstance(value, str):
            return self.sanitize_text(value)

        if value is None or isinstance(value, (bool, int, float)):
            return PiiSanitizationResult(value=value, redactions={})

        seen = _seen if _seen is not None else set()
        value_id = id(value)
        if value_id in seen:
            return PiiSanitizationResult(value="[REDACTED_CIRCULAR_REFERENCE]", redactions={})

        seen.add(value_id)
        try:
            if isinstance(value, list):
                redactions: PiiRedactionCounts = {}
                sanitized_items = []
                for item in value:
                    sanitized = self.sanitize_value(item, seen)
                    redactions = merge_redactions(redactions, sanitized.redactions)
                    sanitized_items.append(sanitized.value)
                return PiiSanitizationResult(value=sanitized_items, redactions=redactions)

            if isinstance(value, tuple):
                redactions = {}
                sanitized_items = []
                for item in value:
                    sanitized = self.sanitize_value(item, seen)
                    redactions = merge_redactions(redactions, sanitized.redactions)
                    sanitized_items.append(sanitized.value)
                return PiiSanitizationResult(value=tuple(sanitized_items), redactions=redactions)

            if isinstance(value, dict):
                redactions = {}
                sanitized_dict = {}
                for key, item in value.items():
                    if isinstance(key, str):
                        sanitized_key = self.sanitize_text(key)
                        safe_key = sanitized_key.value
                        redactions = merge_redactions(redactions, sanitized_key.redactions)
                    else:
                        safe_key = key

                    sanitized_item = self.sanitize_value(item, seen)
                    redactions = merge_redactions(redactions, sanitized_item.redactions)
                    sanitized_dict[safe_key] = sanitized_item.value

                return PiiSanitizationResult(value=sanitized_dict, redactions=redactions)

            return PiiSanitizationResult(value=value, redactions={})
        finally:
            seen.remove(value_id)

    # ── rehydrate ─────────────────────────────────────────────────────────────

    def rehydrate_text(self, text: str) -> str:
        """Substitute known placeholders back to their real values.

        Placeholders this session never issued (e.g. a model-invented
        "[EMAIL_99]") are left as-is rather than guessed at.
        """
        if not self._value_by_placeholder:
            return text

        def replace(match: re.Match[str]) -> str:
            return self._value_by_placeholder.get(match.group(0), match.group(0))

        return _PLACEHOLDER_RE.sub(replace, text)

    def rehydrate_value(self, value: Any, _seen: set[int] | None = None) -> Any:
        if isinstance(value, str):
            return self.rehydrate_text(value)

        if value is None or isinstance(value, (bool, int, float)):
            return value

        seen = _seen if _seen is not None else set()
        value_id = id(value)
        if value_id in seen:
            return value

        seen.add(value_id)
        try:
            if isinstance(value, list):
                return [self.rehydrate_value(item, seen) for item in value]
            if isinstance(value, tuple):
                return tuple(self.rehydrate_value(item, seen) for item in value)
            if isinstance(value, dict):
                return {
                    (self.rehydrate_text(key) if isinstance(key, str) else key): self.rehydrate_value(item, seen)
                    for key, item in value.items()
                }
            return value
        finally:
            seen.remove(value_id)


# ── Stateless wrappers (one-shot, mapping discarded → effectively destructive) ──
#
# Kept for callers that sanitize without a run context. New code with a run
# lifecycle should hold a PseudonymizationSession and rehydrate outputs.


def sanitize_text_for_llm(text: str) -> PiiSanitizationResult:
    return PseudonymizationSession().sanitize_text(text)


def sanitize_value_for_llm(value: Any, _seen: set[int] | None = None) -> PiiSanitizationResult:
    return PseudonymizationSession().sanitize_value(value, _seen)
